demo/popcha-client/web/src/authentication.py

```diff
@@ -0,0 +1,118 @@
+"""
+This file includes all the authentication logic. No external server logic is
+done here.
+"""
+
+# allows class names to be used as return types inside the class
+# see e.g. https://stackoverflow.com/a/61544901
+from __future__ import annotations
+
+import secrets
+import time
+from urllib import parse
+
+import jwt
+from jwt import PyJWTError
+from werkzeug.datastructures import MultiDict
+
+
+class Authentication:
+    """
+    Class that allows users to authenticate
+    """
+
+    def __init__(self, providers: list[dict[str, str]] = None):
+        if providers is None:
+            providers = []
+
+        self.providers: list[dict[str, str]] = providers
+        self.login_states: dict[str, (str, str, float)] = {}
+
+    def get_url(
+            self, auth_server: str, lao_id: str, host_server: str,
+            host_port: str | int, client_id: str
+            ) -> str:
+        """
+        Generates the url to contact the authentication server. It contains all
+        the information required by the authentication server
+        :param auth_server: The domain of the authentication server (OpenID
+        provider)
+        :param lao_id: The LAO ID the user wants to connect with
+        :param host_server: The current host server domain
+        :param host_port: The port the host server is running on
+        :param client_id: The unique identifier of this client
+        :return: A url to the authentication server (Open ID Provider)
+        """
+        nonce = secrets.token_urlsafe(64)
+        state = secrets.token_urlsafe(64)
+
+        self.login_states[state] = (nonce, auth_server, time.time())
+
+        parameters = {
+            "response_mode": "query",
+            "response_type": "id_token",
+            "client_id": client_id,
+            "redirect_uri": f"https://{host_server}:{str(host_port)}/cb",
+            "scope": "openid profile",
+            "login_hint": lao_id,
+            "nonce": nonce,
+            "state": state
+            }
+
+        return f"https://{auth_server}/authorize?{parse.urlencode(parameters)}"
+
+    def validate_args(self, args: MultiDict[str, str], client_id: str) \
+            -> str | None:
+        """
+        Validate the args as if they were coming from OpenID Authentication
+        request
+        :param args: A MultiDict containing the arguments. Similar to a dict
+        but allows multiple values for a key. Implemented as dict[str, list[str]].
+        See https://werkzeug.palletsprojects.com/en/2.3.x/datastructures/#werkzeug.datastructures.MultiDict
+        :param client_id: The ID of this client
+        :return: The unique user identifier if logged in, None otherwise
```

Comment:
i don't know if it is possible to have a specific error returned if there is a problem (instead of a general "None")

Reply:
The idea of this function is to return the ID of the user. I think it is better to return `None` since it is the current user's ID based on the provided arguments.